### ISSUE TITLE
[ECS, Fargate]: Support for running Docker-in-Docker containers

### ISSUE THREAD

**Issue by drissamri** [aws/containers-roadmap#1356]
**Tell us about your request**
Currently it is not possible to use Docker-in-Docker on Fargate. We use Fargate to run our Jenkins jobs, and there are a lot of cases where it is useful to have run Docker inside those jobs. 

One of our examples is running [Testcontainers](https://www.testcontainers.org/) inside our tests to mock external dependencies like DynamoDB, Queues, Redis, ... and so on.

I was hoping the solution would be an outcome of https://github.com/aws/containers-roadmap/issues/95 but this was recently closed with an alternative way to build containers. 

**Which service(s) is this request for?**
Fargate, ECS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by antgel** [comment]
Thanks for opening this, I don't believe that #95 should have been closed.

**Comment by mehtaparas** [comment]
I agree, I don't believe #95 should have been closed. Can we get an update / response on this issue? 

For context, I'm interested in this issue because I want to be able to deploy [Airbyte on ECS](https://github.com/airbytehq/airbyte/issues/4847).

**Comment by nextroll-marcus** [comment]
Really wish I'd realized this before taking the time to re-write an ec2 asg workflow that does nothing but build containers into fargate, the fact you can't do this should be in big blinking text before you end up reading anything else

**Comment by boyapatisandeep** [comment]
Any update on this using docker in docker for Fargate ?

**Comment by mjy90** [comment]
Any updates? This feature would be fantastically useful for our business.

**Comment by nabsource** [comment]
This would be a great addition, would it be possible to get an update on this?

**Comment by aaronsteers** [comment]
As a workaround, has anyone been successful in declaring a sidecar container in the ECS task definition?

A challenge with this approach is that the secondary container's image has to be known ahead of time, and the container needs to wait in a holding pattern until it receives some command or workload from the primary container. Logistics of this are non-trivial, but at least _in theory_ it could/should be possible for certain use cases. 

Curious if anyone has tried this approach or knows ahead if it would be viable.

**Comment by AllanOricil** [comment]
Another use case to use it is with Github Actions when jobs run inside containers.

**Comment by jerryeml** [comment]
looking forward to update this

**Comment by RazvanGherlea** [comment]
Hi All,

Are you saying that if we would like to use the ECS Fargate Service container as self hsoted runner we cannot define a specific container in the ci/cd pipeline like the following example ?

https://docs.gitlab.com/ee/ci/docker/using_docker_images.html#define-image-in-the-gitlab-ciyml-file

Thanks in advance

**Comment by billyjbryant** [comment]
@RazvanGherlea 

> Are you saying that if we would like to use the ECS Fargate Service container as self hsoted runner we cannot define a specific container in the ci/cd pipeline like the following example ?

Yes, the issue currently, is that you cannot run Docker in Docker (DinD) on ECS Fargate as the task runs without `--privileged` mode and will not allow the container launched in your Self Hosted Runner's task to mount or interact with the container's filesystem.

**Comment by OpenSourceable** [comment]
@billyjbryant Have you considered using [Kaniko](https://github.com/GoogleContainerTools/kaniko) as suggested [here](https://aws.amazon.com/blogs/containers/building-container-images-on-amazon-ecs-on-aws-fargate/)? This repo has some examples: https://github.com/int128/kaniko-action

**Comment by jesseadams** [comment]
I'm not sure why everyone keeps reading this issue as "needing to build a docker image in ECS Fargate" versus what its actually looking for, which is running a docker container within a running Task in ECS Fargate.

Example: An ECS Fargate task running `jenkins/inbound-agent:latest-alpine-jdk21` with a Jenkinsfile that tries to run the following.

```
node('ecs') {
  stage('Checkout Code') {
    scm_vars = checkout scm
    env.GIT_COMMIT = scm_vars.GIT_COMMIT
  }

  stage('Node Version') {
      docker.image('node:lts-alpine').inside {
          sh 'node -v'
      }
  }
}
```
Otherwise the jenkins build agent container needs to have EVERYTHING installed on it that you need for all of your pipelines, which is unreasonable.

**Comment by ollypom** [comment]
While it is not possible to run Docker in Docker (DinD) on Amazon ECS with AWS Fargate today, for those building with the new [Amazon ECS Managed Instances](https://aws.amazon.com/ecs/managed-instances/), I have published an example on how to build container images using Docker-in-Docker on an ECS Managed Instances capacity provider [here](https://gist.github.com/ollypom/924e6075cbc9760a216730f6c6833d77).

**Comment by AbhishekNautiyal** [comment]
Thanks for publishing the example, @ollypom!

Customers an now run Docker-in-Docker on Amazon ECS Managed Instances - a fully managed compute option that delivers the same operational and security benefits as Fargate while providing access to the full range of EC2 capabilities.  For more details refer to https://github.com/aws/containers-roadmap/issues/2452

We are aware that this feature request was originally opened for Fargate. While ECS Managed Instances addresses this need, some customers may still want to run Docker-in-Docker containers on Fargate. I am closing this issue now so that we can focus and gather feedback for prioritization of Fargate-specific features. Please open a new issue if you still require DinD support on Fargate, including your use case(s), or if you have any other requests. As always, we appreciate your engagement and feedback on this thread.
